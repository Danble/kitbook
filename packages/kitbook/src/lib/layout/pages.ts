export type Folder = {
  name: string;
  url: string;
  depth: number;
  folders?: Folder[];
  pages?: Page[];
};

export type Page = {
  name: string;
  url: string;
  ext?: string;
  path?: string;
  // organize sibling modules into 1 page, show stories/docs first, then default component view, then variants
  storyModulePath?: string;
  componentModulePath?: string;
  pageModulePath?: string;
  variantsModulePath?: string;
};

export type Modules = Record<string, () => Promise<{ [key: string]: any }>>;

function removeInitialDigitAndHyphens(string: string) {
  return string.replace(/^\d+/, '').replace(/-/g, ' ').trim();
}

export function parsePath(path: string) {
  if (path === '/README.md') return { ext: 'md', name: 'README', dir: '/' }

  const match = path.match(/^\/src\/(.*\/)[+_]?(.+?)\.(.+)$/);
  if (!match) throw new Error(`${path} is not a module path that Kitbook can handle. Make sure your Kitbook Layout Load import meta glob starts with '/src/**'`);
  const [, dir, name, ext] = match;
  return { dir, name, ext };
}

export function parseModules(modules: Modules, root = '/kitbook'): Page[] {
  const paths = Object.keys(modules)
    .filter(p => {
      return !p.includes(`/src/routes${root}`) &&
        !p.includes('+layout.svelte')
    });

  if (!paths.length) return []

  const pages = paths.map((path) => {
    const { name, ext } = parsePath(path);
    const url = path.replace('src/', '').replace(`.${ext}`, '');
    return {
      path,
      name: removeInitialDigitAndHyphens(name),
      ext,
      url,
    };
  });
  return pages;
}

export function combineModulesIntoPages(uncombined: Page[]): Page[] {
  const combined: Record<string, Page> = {};

  for (const page of uncombined) {
    const url = page.url.replace(/[+_]?page/, 'page'); // strip page prefixes for merging
    if (!combined[url])
      combined[url] = { name: page.name, url: page.url }

    // Will skip files not matching expected extensions, e.g. /src/A/Bar.foo.svelte
    if (['md', 'svx'].includes(page.ext)) {
      combined[url].storyModulePath = page.path
    } else if (page.name === '+page') {
      combined[url].pageModulePath = page.path
    } else if (page.ext === 'svelte') {
      combined[url].componentModulePath = page.path
    } else if (page.ext === 'variants.ts') {
      combined[url].variantsModulePath = page.path
    }
  }

  return Object.values(combined);
}

export function putPagesIntoFolders(pagesToOrganize: Page[]): Folder {
  const rootFolder: Folder = {
    name: '.',
    url: '/',
    depth: 0,
    folders: [],
    pages: [],
  };

  if (!pagesToOrganize?.length) return rootFolder;

  pagesToOrganize.forEach((page) => {
    let currentFolder = rootFolder;

    const path = page.url.split('/');
    path.forEach((folderName, index) => {
      const cleanedFolderName = removeInitialDigitAndHyphens(folderName);

      if (index === path.length - 1) {
        currentFolder.pages.push(page); // is a page
      } else if (folderName === '') {
        currentFolder = rootFolder;
      } else {
        let folder = currentFolder.folders.find((folder) => folder.name === cleanedFolderName);
        if (!folder) {
          folder = {
            name: cleanedFolderName,
            url: path
              .slice(0, index + 1)
              .join('/')
              .replace(/^\./, ''),
            depth: index,
            folders: [],
            pages: [],
          };
          currentFolder.folders.push(folder);
        }
        currentFolder = folder;
      }
    });
  });
  return rootFolder;
}


