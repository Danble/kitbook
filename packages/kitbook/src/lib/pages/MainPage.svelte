<script lang="ts">
  import { page } from '$app/stores';
  import { compressToEncodedURIComponent as encode } from 'lz-string';
  import type { GroupedPage, GroupedPageMap, LoadedModules } from '../kitbook-types';
  import EditInGithub from '../components/EditInGithub.svelte';

  export let data: {
    pages?: GroupedPageMap;
    page?: GroupedPage;
    loadedModules?: LoadedModules;
    error?: string;
  } = { loadedModules: {} };

  $: pathWouldRecurseInfinitelyIfInSandbox = $page.url.pathname.startsWith(
    '/lib/routes/sandbox/[...file]/+'
  );
  $: doesNotHaveSvxOrVariants = !(data.loadedModules.svx || data.loadedModules.variants);
</script>

{#if data.error}
  {data.error}
{:else}
  {#if data.loadedModules.svx}
    <div class="mb-10">
      <svelte:component this={data.loadedModules.svx} />
    </div>
  {/if}

  {#if data.loadedModules.component}
    {#if data.loadedModules.variants}
      <div class="text-2xl">
        {data.page.name.startsWith('+') ? 'Page' : 'Component'} Variants
      </div>
      {#each data.loadedModules.variants as variant, index}
        <div class="not-prose border rounded mt-3">
          <div class="bg-gray-200 p-3 mb-2">
            <span class="font-semibold">
              {variant.name}
            </span>
            {#if variant.description}
              <div class="text-sm">
                {variant.description}
              </div>
            {/if}
          </div>
          {#if pathWouldRecurseInfinitelyIfInSandbox}
            <svelte:component this={data.loadedModules.component} {...variant.props || {}} />
          {:else}
            <iframe
              class="w-full h-full"
              title=""
              src="/sandbox/{$page.url.pathname}?props={encode(
                JSON.stringify(variant.props || {})
              )}&variantIdx={index}"
            />
          {/if}
        </div>
      {/each}
      <!-- {:else if !pathWouldRecurseInfinitelyIfInSandbox}
      <div class="not-prose border rounded mt-3">
        <div class="bg-gray-200 p-3 mb-2">
          <span class="font-semibold"> Default </span>
        </div>
        <iframe
          class="w-full h-full"
          title=""
          src="/sandbox/{$page.url.pathname}?props={encode(JSON.stringify({}))}"
        />
      </div> -->
    {/if}
  {/if}

  {#if doesNotHaveSvxOrVariants}
    <div class="mb-3 p-3 bg-gray-200 rounded">
      <b>Kitbook tip</b>: You have not created a Stories file ({data.page?.name}.svx/.md) nor a Variants file ({data
        .page?.name}.variants.ts) file. In the future Kitbook will try to
      automatically supply default props, but until then you must manually scaffold a page for this
      component by creating either a Stories or Variants file.
    </div>
  {/if}

  <EditInGithub path={data?.page?.path} />
{/if}
