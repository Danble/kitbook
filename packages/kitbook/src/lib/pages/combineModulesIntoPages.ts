import type { Page, PageMap } from "$lib/kitbook-types";
import { testModules } from "./testModules";
import { parseModules } from "./parseModules";

/** allow _page/_layout to be merged with +page/+layout */
function convertUnderscorePrefixToPlus(s: string): string {
  return s.replace('_page', '+page').replace('_layout', '+layout')
}

export function combineModulesIntoPages(uncombined: Page[]): PageMap {
  const combined: PageMap = {};

  for (const page of uncombined) {
    const url = convertUnderscorePrefixToPlus(page.url);
    if (!combined[url])
      combined[url] = { name: page.name, url: page.url }

    // Will skip files not matching expected extensions, e.g. /src/A/Bar.foo.svelte
    if (['md', 'svx'].includes(page.ext)) {
      combined[url].svxModulePath = page.path
    } else if ((page.name === '+page') || (page.name === '+layout')) {
      combined[url].pageModulePath = page.path // layouts are just pages with slot inheritance super-powers
    } else if (page.ext === 'svelte') {
      combined[url].componentModulePath = page.path
    } else if (page.ext === 'variants.ts') {
      combined[url].variantsModulePath = page.path
    }
  }

  return combined;
}

if (import.meta.vitest) {
  test('combineModulesIntoPages properly combines Pages', () => {
    const pages = parseModules(testModules);
    expect(combineModulesIntoPages(pages)).toMatchInlineSnapshot(`
      {
        "/README": {
          "name": "README",
          "svxModulePath": "/README.md",
          "url": "/README",
        },
        "/docs/0-why-kitbook": {
          "name": "why kitbook",
          "svxModulePath": "/src/docs/0-why-kitbook.md",
          "url": "/docs/0-why-kitbook",
        },
        "/docs/1-get-started": {
          "name": "get started",
          "svxModulePath": "/src/docs/1-get-started.md",
          "url": "/docs/1-get-started",
        },
        "/docs/index": {
          "name": "index",
          "svxModulePath": "/src/docs/index.md",
          "url": "/docs/index",
        },
        "/docs/my-notes/0-unocss": {
          "name": "unocss",
          "svxModulePath": "/src/docs/my-notes/0-unocss.md",
          "url": "/docs/my-notes/0-unocss",
        },
        "/docs/my-notes/1-deploy-to-vercel": {
          "name": "deploy to vercel",
          "svxModulePath": "/src/docs/my-notes/1-deploy-to-vercel.md",
          "url": "/docs/my-notes/1-deploy-to-vercel",
        },
        "/lib/A": {
          "componentModulePath": "/src/lib/A.svelte",
          "name": "A",
          "url": "/lib/A",
        },
        "/lib/B": {
          "componentModulePath": "/src/lib/B.svelte",
          "name": "B",
          "svxModulePath": "/src/lib/B.svx",
          "url": "/lib/B",
        },
        "/lib/E": {
          "name": "E",
          "svxModulePath": "/src/lib/E.svx",
          "url": "/lib/E",
        },
        "/lib/a/C": {
          "componentModulePath": "/src/lib/a/C.svelte",
          "name": "C",
          "url": "/lib/a/C",
          "variantsModulePath": "/src/lib/a/C.variants.ts",
        },
        "/lib/a/D": {
          "componentModulePath": "/src/lib/a/D.svelte",
          "name": "D",
          "svxModulePath": "/src/lib/a/D.svx",
          "url": "/lib/a/D",
          "variantsModulePath": "/src/lib/a/D.variants.ts",
        },
        "/routes/+layout": {
          "name": "+layout",
          "pageModulePath": "/src/routes/+layout.svelte",
          "url": "/routes/+layout",
        },
        "/routes/+page": {
          "name": "+page",
          "pageModulePath": "/src/routes/+page.svelte",
          "url": "/routes/+page",
        },
        "/routes/a/+page": {
          "name": "+page",
          "pageModulePath": "/src/routes/a/+page.svelte",
          "svxModulePath": "/src/routes/a/_page.svx",
          "url": "/routes/a/+page",
        },
        "/routes/b/+page": {
          "name": "+page",
          "pageModulePath": "/src/routes/b/+page.svelte",
          "url": "/routes/b/+page",
          "variantsModulePath": "/src/routes/b/_page.variants.ts",
        },
        "/routes/c/+page": {
          "name": "+page",
          "pageModulePath": "/src/routes/c/+page.svelte",
          "svxModulePath": "/src/routes/c/_page.svx",
          "url": "/routes/c/+page",
          "variantsModulePath": "/src/routes/c/_page.variants.ts",
        },
      }
    `);
  });
}