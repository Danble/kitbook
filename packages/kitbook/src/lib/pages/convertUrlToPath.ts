export function convertUrlToModulePath(url: string, extension: 'md' | 'svx' | 'svelte' = 'svelte'): string {
  let path = extension === 'svelte' ? url : url.replace('/+page', '/_page');
  return `/src${path}.${extension}`;
}