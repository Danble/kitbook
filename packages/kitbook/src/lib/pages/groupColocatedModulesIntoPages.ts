import type { Modules, RawModules, GroupedPageMap } from "../kitbook-types";
import { groupColocatedPages } from "./groupColocatedPages";
import { parseModulesIntoUngroupedPages } from "./parseModulesIntoUngroupedPages";
import { testModules } from "./testModules";

export function groupColocatedModulesIntoPages(modules: Modules, modulesRaw: RawModules): GroupedPageMap {
  const ungroupedPages = parseModulesIntoUngroupedPages(modules, modulesRaw);
  return groupColocatedPages(ungroupedPages);
}

if (import.meta.vitest) {
  // this test is redundant with the one in groupColocatedPages.ts but it's here to make sure that the two functions work together (would be good to figure out how to remove it)
  test('groupColocatedModulesIntoPages', () => {
    expect(groupColocatedModulesIntoPages(testModules, testModules)).toMatchInlineSnapshot(`
      {
        "/README": {
          "extensions": [
            "md",
          ],
          "loadSvx": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "README",
          "path": "/README.md",
          "url": "/README",
        },
        "/docs/0-why-kitbook": {
          "extensions": [
            "md",
          ],
          "loadSvx": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "why kitbook",
          "path": "/src/docs/0-why-kitbook.md",
          "url": "/docs/0-why-kitbook",
        },
        "/docs/1-get-started": {
          "extensions": [
            "md",
          ],
          "loadSvx": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "get started",
          "path": "/src/docs/1-get-started.md",
          "url": "/docs/1-get-started",
        },
        "/docs/my-notes/0-unocss": {
          "extensions": [
            "md",
          ],
          "loadSvx": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "unocss",
          "path": "/src/docs/my-notes/0-unocss.md",
          "url": "/docs/my-notes/0-unocss",
        },
        "/docs/my-notes/1-deploy-to-vercel": {
          "extensions": [
            "md",
          ],
          "loadSvx": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "deploy to vercel",
          "path": "/src/docs/my-notes/1-deploy-to-vercel.md",
          "url": "/docs/my-notes/1-deploy-to-vercel",
        },
        "/index": {
          "extensions": [
            "md",
          ],
          "loadSvx": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "index",
          "path": "/src/index.md",
          "url": "/index",
        },
        "/kitbook/(main)/+layout": {
          "extensions": [
            "svelte",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "+layout",
          "path": "/src/kitbook/(main)/+layout.svelte",
          "url": "/kitbook/(main)/+layout",
        },
        "/kitbook/(main)/[...file]/+page": {
          "extensions": [
            "svelte",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "+page",
          "path": "/src/kitbook/(main)/[...file]/+page.svelte",
          "url": "/kitbook/(main)/[...file]/+page",
        },
        "/kitbook/sandbox/[...file]/+page": {
          "extensions": [
            "svelte",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "+page",
          "path": "/src/kitbook/sandbox/[...file]/+page.svelte",
          "url": "/kitbook/sandbox/[...file]/+page",
        },
        "/lib/A": {
          "extensions": [
            "svelte",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "A",
          "path": "/src/lib/A.svelte",
          "url": "/lib/A",
        },
        "/lib/B": {
          "extensions": [
            "svelte",
            "svx",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "loadSvx": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "B",
          "path": "/src/lib/B.svelte",
          "url": "/lib/B",
        },
        "/lib/E": {
          "extensions": [
            "svx",
          ],
          "loadSvx": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "E",
          "path": "/src/lib/E.svx",
          "url": "/lib/E",
        },
        "/lib/a/C": {
          "extensions": [
            "svelte",
            "variants.ts",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "loadVariants": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "C",
          "path": "/src/lib/a/C.svelte",
          "url": "/lib/a/C",
        },
        "/lib/a/D": {
          "extensions": [
            "svelte",
            "svx",
            "variants.ts",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "loadSvx": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "loadVariants": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "D",
          "path": "/src/lib/a/D.svelte",
          "url": "/lib/a/D",
        },
        "/routes/+layout": {
          "extensions": [
            "svelte",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "+layout",
          "path": "/src/routes/+layout.svelte",
          "url": "/routes/+layout",
        },
        "/routes/+page": {
          "extensions": [
            "svelte",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "+page",
          "path": "/src/routes/+page.svelte",
          "url": "/routes/+page",
        },
        "/routes/a/+layout": {
          "extensions": [
            "svelte",
            "svx",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "loadSvx": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "+layout",
          "path": "/src/routes/a/+layout.svelte",
          "url": "/routes/a/+layout",
        },
        "/routes/a/+page": {
          "extensions": [
            "svelte",
            "svx",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "loadSvx": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "+page",
          "path": "/src/routes/a/+page.svelte",
          "url": "/routes/a/+page",
        },
        "/routes/b/+layout": {
          "extensions": [
            "svelte",
            "variants.ts",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "loadVariants": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "+layout",
          "path": "/src/routes/b/+layout.svelte",
          "url": "/routes/b/+layout",
        },
        "/routes/b/+page": {
          "extensions": [
            "svelte",
            "variants.ts",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "loadVariants": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "+page",
          "path": "/src/routes/b/+page.svelte",
          "url": "/routes/b/+page",
        },
        "/routes/c/+layout": {
          "extensions": [
            "svelte",
            "svx",
            "variants.ts",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "loadSvx": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "loadVariants": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "+layout",
          "path": "/src/routes/c/+layout.svelte",
          "url": "/routes/c/+layout",
        },
        "/routes/c/+page": {
          "extensions": [
            "svelte",
            "svx",
            "variants.ts",
          ],
          "loadComponent": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "loadSvx": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "loadVariants": {
            "loadModule": [Function],
            "loadRaw": [Function],
          },
          "name": "+page",
          "path": "/src/routes/c/+page.svelte",
          "url": "/routes/c/+page",
        },
      }
    `);
  });
}