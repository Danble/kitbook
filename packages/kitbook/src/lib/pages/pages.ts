import type { Folder, Page, Modules } from "kitbook";

type PageMap = Record<string, Page>;

function removeInitialDigitAndHyphens(string: string) {
  return string.replace(/^\d+/, '').replace(/-/g, ' ').trim();
}
/** allow _page/_layout to be merged with +page/+layout */
function convertUnderscorePrefixToPlus(s: string): string {
  return s.replace('_page', '+page').replace('_layout', '+layout')
}

export function parsePath(path: string) {
  if (path === '/README.md') return { ext: 'md', name: 'README', dir: '/' }

  const match = path.match(/^\/src\/(.*\/)(.+?)\.(.+)$/);
  if (!match) throw new Error(`${path} is not a module path that Kitbook can handle. Make sure your Kitbook Layout Load import meta glob starts with '/src/**'`);
  const [, dir, name, ext] = match;
  return { dir, name, ext };
}

export function parseModules(modules: Modules): Page[] {
  const paths = Object.keys(modules)
    .filter(p => {
      return !p.includes(`/src/kitbook`) &&
        !p.includes('+layout.svelte')
    });

  if (!paths.length) return []

  const pages = paths.map((path) => {
    const { name, ext } = parsePath(path);
    const url = path.replace('src/', '').replace(`.${ext}`, '');
    return {
      path,
      name: removeInitialDigitAndHyphens(name),
      ext,
      url,
    };
  });
  return pages;
}

export function combineModulesIntoPages(uncombined: Page[]): PageMap {
  const combined: PageMap = {};

  for (const page of uncombined) {
    const url = convertUnderscorePrefixToPlus(page.url);
    if (!combined[url])
      combined[url] = { name: page.name, url: page.url }

    // Will skip files not matching expected extensions, e.g. /src/A/Bar.foo.svelte
    if (['md', 'svx'].includes(page.ext)) {
      combined[url].svxModulePath = page.path
    } else if (page.name === '+page' || page.name === '+layout') {
      combined[url].pageModulePath = page.path // layouts are just pages with slot inheritance super-powers
    } else if (page.ext === 'svelte') {
      combined[url].componentModulePath = page.path
    } else if (page.ext === 'variants.ts') {
      combined[url].variantsModulePath = page.path
    }
  }

  return combined;
}

export function putPagesIntoFolders(combinedPages: PageMap): Folder {
  const rootFolder: Folder = {
    name: '.',
    url: '/',
    depth: 0,
    folders: [],
    pages: [],
  };

  const pagesToOrganize = Object.values(combinedPages)
  if (!pagesToOrganize?.length) return rootFolder;

  pagesToOrganize.forEach((page) => {
    let currentFolder = rootFolder;

    const path = page.url.split('/');
    path.forEach((folderName, index) => {
      const cleanedFolderName = removeInitialDigitAndHyphens(folderName);

      if (index === path.length - 1) {
        currentFolder.pages.push(page); // is a page
      } else if (folderName === '') {
        currentFolder = rootFolder;
      } else {
        let folder = currentFolder.folders.find((folder) => folder.name === cleanedFolderName);
        if (!folder) {
          folder = {
            name: cleanedFolderName,
            url: path
              .slice(0, index + 1)
              .join('/')
              .replace(/^\./, ''),
            depth: index,
            folders: [],
            pages: [],
          };
          currentFolder.folders.push(folder);
        }
        currentFolder = folder;
      }
    });
  });
  return rootFolder;
}


