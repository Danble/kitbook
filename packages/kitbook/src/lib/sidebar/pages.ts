export type Folder = {
  name: string;
  url: string;
  depth: number;
  folders?: Folder[];
  pages?: Page[];
};

export type Page = {
  name: string;
  ext: string;
  path: string;
  url: string;
};

export type Modules = Record<string, () => Promise<{ [key: string]: any }>>;

function removeInitialDigitAndHyphens(string: string) {
  return string.replace(/^\d+/, '').replace(/-/g, ' ').trim();
}

export function parsePath(path: string) {
  if (path === '/README.md') return { ext: 'md', name: 'README', dir: '/' }

  const match = path.match(/^\/src\/(.*\/)(.+?)\.(.+)$/);
  if (!match) throw new Error(`${path} is not a module path that Kitbook can handle. Make sure your Kitbook Layout Load import meta glob starts with '/src/**'`);
  const [, dir, name, ext] = match;
  return { dir, name, ext };
}

export function parseModules(modules: Modules, root = '/kitbook'): Page[] {
  const paths = Object.keys(modules)
    .filter(p => {
      return !p.includes(`/src/routes${root}`) &&
        !p.includes('+layout.svelte')
    });

  if (!paths.length) return []

  const pages = paths.map((path) => {
    const { dir, name, ext } = parsePath(path);
    return {
      path: path.replace('/src/', ''),
      name: removeInitialDigitAndHyphens(name),
      ext,
      url: dir,
    };
  });
  return pages;
}

export function putPagesIntoFolders(pagesToOrganize: Page[]): Folder {
  const rootFolder: Folder = {
    name: '.',
    url: '/',
    depth: 0,
    folders: [],
    pages: [],
  };

  if (!pagesToOrganize?.length) return rootFolder;

  pagesToOrganize.forEach((page) => {
    let currentFolder = rootFolder;

    const path = page.url.split('/');
    path.forEach((folderName, index) => {
      const cleanedFolderName = removeInitialDigitAndHyphens(folderName);

      if (index === path.length - 1) {
        currentFolder.pages.push(page); // is a page
      } else if (folderName === '') {
        currentFolder = rootFolder;
      } else {
        let folder = currentFolder.folders.find((folder) => folder.name === cleanedFolderName);
        if (!folder) {
          folder = {
            name: cleanedFolderName,
            url: path
              .slice(0, index + 1)
              .join('/')
              .replace(/^\./, ''),
            depth: index + 1,
            folders: [],
            pages: [],
          };
          currentFolder.folders.push(folder);
        }
        currentFolder = folder;
      }
    });
  });
  return rootFolder;
}

// export function findActivePage(pages: Page[], urlPathname: string): Page {
//   const activePage = pages.find((page) => {
//     const regex = new RegExp(`${page.url}$`);
//     return regex.test(urlPathname);
//   });
//   if (activePage) return activePage;
//   return pages.find((page) => page.url === '/');
// }
