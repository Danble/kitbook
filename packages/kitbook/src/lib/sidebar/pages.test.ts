import { findActivePage, parsePages, putPagesIntoFolders } from "./pages";

const modules = {
  // ./0-components/0-layout.md
  // ./0-components/1-story.svx
  // ./0-components/internal/0-sidebar.svx
  // ./0-components/internal/1-knobs.svx
  // ./0-why-kitbook.md
  // ./1-get-started.md
  // ./2-add-stories.md
  // ./9-maintainer-notes/0-add-windicss.md
  // ./9-maintainer-notes/1-deploy-to-vercel.md
  './9-privacy-policy.md': () => Promise.resolve({}),
  './0-get-started.md': () => Promise.resolve({}),
  './0-components/0-Button.svelte': () => Promise.resolve({}),
  './0-components/1-Switch.svelte': () => Promise.resolve({}),
  './0-components/0-ui/0-Button.svelte': () => Promise.resolve({}),
  './0-components/play-audio-section.svelte': () => Promise.resolve({}), // test this to remove section hyphen
  './3-examples.md': () => Promise.resolve({}),
  './[reference]/__layout.svelte': () => Promise.resolve({}),
  './a/b/c-d/e.svelte': () => Promise.resolve({}),
  './a/b/c-d/f.svelte': () => Promise.resolve({}),
  './index.md': () => Promise.resolve({}),
};
const pages = parsePages(modules);

test('putPagesIntoFolders organizes pages into proper folders based on dir', () => {
  expect(putPagesIntoFolders(pages)).toMatchSnapshot();
});

test('parsePages properly returns array of Page objects', () => {
  expect(parsePages(modules)).toMatchInlineSnapshot(`
      [
        {
          "ext": "md",
          "name": "index",
          "path": "./index.md",
          "url": "/",
        },
        {
          "ext": "md",
          "name": "privacy policy",
          "path": "./9-privacy-policy.md",
          "url": "/9-privacy-policy",
        },
        {
          "ext": "md",
          "name": "get started",
          "path": "./0-get-started.md",
          "url": "/0-get-started",
        },
        {
          "ext": "svelte",
          "name": "Button",
          "path": "./0-components/0-Button.svelte",
          "url": "/0-components/0-Button",
        },
        {
          "ext": "svelte",
          "name": "Switch",
          "path": "./0-components/1-Switch.svelte",
          "url": "/0-components/1-Switch",
        },
        {
          "ext": "svelte",
          "name": "Button",
          "path": "./0-components/0-ui/0-Button.svelte",
          "url": "/0-components/0-ui/0-Button",
        },
        {
          "ext": "svelte",
          "name": "play audio section",
          "path": "./0-components/play-audio-section.svelte",
          "url": "/0-components/play-audio-section",
        },
        {
          "ext": "md",
          "name": "examples",
          "path": "./3-examples.md",
          "url": "/3-examples",
        },
        {
          "ext": "svelte",
          "name": "e",
          "path": "./a/b/c-d/e.svelte",
          "url": "/a/b/c-d/e",
        },
        {
          "ext": "svelte",
          "name": "f",
          "path": "./a/b/c-d/f.svelte",
          "url": "/a/b/c-d/f",
        },
      ]
    `);
});

test('findActivePage returns page that includes current url', () => {
  expect(findActivePage(pages, '/3-examples')).toMatchInlineSnapshot(`
      {
        "ext": "md",
        "name": "examples",
        "path": "./3-examples.md",
        "url": "/3-examples",
      }
    `);
});

test('findActivePage handles index page', () => {
  expect(findActivePage(pages, '/')).toMatchInlineSnapshot(`
      {
        "ext": "md",
        "name": "index",
        "path": "./index.md",
        "url": "/",
      }
    `);
});

test('findActivePage works when kitbook is placed in a subfolder', () => {
  expect(findActivePage(pages, '/foo/kitbookroot/3-examples')).toMatchInlineSnapshot(`
      {
        "ext": "md",
        "name": "examples",
        "path": "./3-examples.md",
        "url": "/3-examples",
      }
    `);
});

test('findActivePage returns kitbook index when no matches', () => {
  expect(findActivePage(pages, '/foo/kitbookroot')).toMatchInlineSnapshot(`
      {
        "ext": "md",
        "name": "index",
        "path": "./index.md",
        "url": "/",
      }
    `);
});
