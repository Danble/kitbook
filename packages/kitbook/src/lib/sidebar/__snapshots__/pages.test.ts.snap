// Vitest Snapshot v1

exports[`putPagesIntoFolders organizes pages into proper folders 1`] = `
{
  "depth": 0,
  "folders": [
    {
      "depth": 1,
      "folders": [
        {
          "depth": 2,
          "folders": [],
          "name": "maintainer notes",
          "pages": [
            {
              "ext": "md",
              "name": "unocss",
              "path": "docs/maintainer-notes/0-unocss.md",
              "url": "docs/maintainer-notes/",
            },
            {
              "ext": "md",
              "name": "deploy to vercel",
              "path": "docs/maintainer-notes/1-deploy-to-vercel.md",
              "url": "docs/maintainer-notes/",
            },
          ],
          "url": "docs/maintainer-notes",
        },
      ],
      "name": "docs",
      "pages": [
        {
          "ext": "md",
          "name": "why kitbook",
          "path": "docs/0-why-kitbook.md",
          "url": "docs/",
        },
        {
          "ext": "md",
          "name": "get started",
          "path": "docs/1-get-started.md",
          "url": "docs/",
        },
        {
          "ext": "md",
          "name": "index",
          "path": "docs/index.md",
          "url": "docs/",
        },
      ],
      "url": "docs",
    },
    {
      "depth": 1,
      "folders": [
        {
          "depth": 2,
          "folders": [],
          "name": "sidebar",
          "pages": [
            {
              "ext": "svelte",
              "name": "Folder",
              "path": "lib/sidebar/Folder.svelte",
              "url": "lib/sidebar/",
            },
            {
              "ext": "variants.ts",
              "name": "Folder",
              "path": "lib/sidebar/Folder.variants.ts",
              "url": "lib/sidebar/",
            },
            {
              "ext": "svelte",
              "name": "Sidebar",
              "path": "lib/sidebar/Sidebar.svelte",
              "url": "lib/sidebar/",
            },
            {
              "ext": "svx",
              "name": "Sidebar",
              "path": "lib/sidebar/Sidebar.svx",
              "url": "lib/sidebar/",
            },
            {
              "ext": "variants.ts",
              "name": "Sidebar",
              "path": "lib/sidebar/Sidebar.variants.ts",
              "url": "lib/sidebar/",
            },
            {
              "ext": "svelte",
              "name": "Page",
              "path": "lib/sidebar/Page.svelte",
              "url": "lib/sidebar/",
            },
          ],
          "url": "lib/sidebar",
        },
      ],
      "name": "lib",
      "pages": [
        {
          "ext": "svelte",
          "name": "Header",
          "path": "lib/Header.svelte",
          "url": "lib/",
        },
        {
          "ext": "svelte",
          "name": "Layout",
          "path": "lib/Layout.svelte",
          "url": "lib/",
        },
        {
          "ext": "svx",
          "name": "Layout",
          "path": "lib/Layout.svx",
          "url": "lib/",
        },
      ],
      "url": "lib",
    },
    {
      "depth": 1,
      "folders": [
        {
          "depth": 2,
          "folders": [],
          "name": "a",
          "pages": [
            {
              "ext": "svelte",
              "name": "+page",
              "path": "routes/a/+page.svelte",
              "url": "routes/a/",
            },
            {
              "ext": "variants.ts",
              "name": "_page",
              "path": "routes/a/_page.variants.ts",
              "url": "routes/a/",
            },
          ],
          "url": "routes/a",
        },
      ],
      "name": "routes",
      "pages": [
        {
          "ext": "svelte",
          "name": "+page",
          "path": "routes/+page.svelte",
          "url": "routes/",
        },
        {
          "ext": "variants.ts",
          "name": "_page",
          "path": "routes/_page.variants.ts",
          "url": "routes/",
        },
      ],
      "url": "routes",
    },
  ],
  "name": ".",
  "pages": [
    {
      "ext": "md",
      "name": "README",
      "path": "/README.md",
      "url": "/",
    },
  ],
  "url": "/",
}
`;
